File: LeetCode/Medium/3201-Find_The_Maximum_Length_Of_Valid_Subsequence_I/3201-FindTheMaximumLengthOfValidSubsequenceI.py
class Solution:
    def maximumLength(self, nums: list[int]) -> int:
        odd = 0
        even = 0
        alternate_even = 0
        alternate_odd = 0
        next_even = True
        next_odd = True
        for num in nums:
            if num & 1:
                # Odd
                odd += 1
                if next_odd:
                    alternate_odd += 1
                    next_odd = False

                if not next_even:
                    alternate_even += 1
                    next_even = True
            else:
                # Even
                even += 1

                if next_even:
                    alternate_even += 1
                    next_even = False

                if not next_odd:
                    alternate_odd += 1
                    next_odd = True

        return max(even, odd, alternate_even, alternate_odd)


a = Solution()
print(a.maximumLength(nums=[1, 2, 3, 4]))
print(a.maximumLength(nums=[1, 2, 1, 1, 2, 1, 2]))
print(a.maximumLength(nums=[1, 3]))

File: LeetCode/Medium/3201-Find_The_Maximum_Length_Of_Valid_Subsequence_I/3201-FindTheMaximumLengthOfValidSubsequenceI.cpp
#include <vector>

class Solution
{
public:
    int maximumLength(vector<int> &nums)
    {
        int odd = 0;
        int even = 0;
        int alternate_even = 0;
        int alternate_odd = 0;
        bool next_even = true;
        bool next_odd = true;

        for (auto num : nums)
        {
            if (num & 1)
            {
                // Odd
                odd += 1;
                if (next_odd)
                {
                    alternate_odd += 1;
                    next_odd = false;
                }

                if (!next_even)
                {
                    alternate_even += 1;
                    next_even = true;
                }
            }
            else
            {
                // Even
                even += 1;

                if (next_even)
                {
                    alternate_even += 1;
                    next_even = false;
                }

                if (!next_odd)
                {
                    alternate_odd += 1;
                    next_odd = true;
                }
            }
        }

        return max({even, odd, alternate_even, alternate_odd});
    }
};